### PR TITLE
Add SGLang CUDA crash API logging inspired by FlashInfer

### PR DESCRIPTION
**Pull request by BBuf**
## Motivation

This PR adds SGLang-native API-level CUDA crash logging for LLM and diffusion kernel call boundaries.

The implementation is inspired by FlashInfer's API logging utility:
https://github.com/flashinfer-ai/flashinfer/blob/main/flashinfer/api_logging.py

This version keeps the scope focused on crash debugging and level-10 dump capture. Replay-related code was intentionally not included so the implementation stays smaller and aligned with the actual SGLang debugging workflow.

Also add a skill to debug cuda crash like flashinfer.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/jit_kernel/debug_utils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import os
+from typing import Any, Callable
+
+
+def maybe_wrap_jit_kernel_sglang_debug(
```
Comment (DarkSharpness) [inline python/sglang/jit_kernel/debug_utils.py]: Could we make it a typing-friendly decorator (like `register_custom_op`)? Also, when used with custom op registration, we need to keep the signature information of the wrapped function.

Comment (BBuf) [reply python/sglang/jit_kernel/debug_utils.py]: Done in https://github.com/sgl-project/sglang/pull/20910/changes/a3af453081af8e9934fce3f10b50c000c636edfe

File: docs/references/environment_variables.md [hunk 2]
```diff
@@ -151,6 +151,11 @@ SGLang supports various environment variables that can be used to configure its
 | `SGLANG_TEST_RETRACT_NO_PREFILL_BS` | When SGLANG_TEST_RETRACT is enabled, no prefill is performed if the batch size exceeds SGLANG_TEST_RETRACT_NO_PREFILL_BS. | `2 ** 31`     |
 | `SGLANG_RECORD_STEP_TIME` | Record step time for profiling | `false` |
 | `SGLANG_TEST_REQUEST_TIME_STATS` | Test request time statistics | `false` |
+| `SGLANG_API_LOGLEVEL` | Controls crash-debug API logging. `0` disables logging, `1` logs API names, `3` logs tensor metadata, `5` adds tensor statistics, and `10` also writes pre-call dump snapshots. | `0` |
```
Comment (merrymercy) [inline docs/references/environment_variables.md]: SGLANG API is too broad. Can you rename all env vars to
`SGLANG_CUDA_API_LOGLEVEL` or `SGLANG_KERNEL_API_LOGLEVEL`?

File: python/sglang/api_logging.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
+"""API-level crash debugging helpers for SGLang.
```
Comment (merrymercy) [inline python/sglang/api_logging.py]: It is okay to call it "api_logging" for flashinfer because the api of flashinfer is all kernels.
It is not very precise to call them "api" of sglang. The API of sglang is "generate", "chat".

File: python/sglang/srt/layers/quantization/bitsandbytes.py [hunk 4]
```diff
@@ -7,6 +7,7 @@
 import torch
 from packaging import version
 
+from sglang.api_logging import sglang_debug_torch_op
```
Comment (merrymercy) [inline python/sglang/srt/layers/quantization/bitsandbytes.py]: this name is too long. make it
```
debug_torch_op
```

File: python/sglang/srt/models/kimi_vl_moonvit.py [hunk 5]
```diff
@@ -65,6 +67,7 @@
 from sglang.srt.utils import add_prefix
 
 
+@sglang_debug_api(op_name="KimiVLMoonViT.multihead_attention")
```
Comment (merrymercy) [inline python/sglang/srt/models/kimi_vl_moonvit.py]: `debug_kernel_api`?

File: sgl-kernel/python/sgl_kernel/flash_attn.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 from typing import Optional, Union
 
 import torch
+from sgl_kernel.debug_utils import maybe_wrap_sglang_debug
```
Comment (merrymercy) [inline sgl-kernel/python/sgl_kernel/flash_attn.py]: `maybe_wrap_debug_kernel`

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: docs/references/environment_variables.md (hunk 2 above)
Comment (BBuf) [reply docs/references/environment_variables.md]: done

File: python/sglang/api_logging.py (hunk 3 above)
Comment (BBuf) [reply python/sglang/api_logging.py]: done

File: python/sglang/srt/layers/quantization/bitsandbytes.py (hunk 4 above)
Comment (BBuf) [reply python/sglang/srt/layers/quantization/bitsandbytes.py]: done

File: python/sglang/srt/models/kimi_vl_moonvit.py (hunk 5 above)
Comment (BBuf) [reply python/sglang/srt/models/kimi_vl_moonvit.py]: done

File: sgl-kernel/python/sgl_kernel/flash_attn.py (hunk 6 above)
Comment (BBuf) [reply sgl-kernel/python/sgl_kernel/flash_attn.py]: done

File: .claude/skills/debug-cuda-crash/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,657 @@
+---
+name: debug-cuda-crash
+description: Tutorial for debugging CUDA crashes in SGLang using kernel API logging
```
Comment (merrymercy) [inline .claude/skills/debug-cuda-crash/SKILL.md]: This should be telling when to call the skill.
e.g.

```
description: Call this skill when you need debug CUDA crashes using kernel API logging.
```

File: python/sglang/jit_kernel/diffusion/triton/norm.py [hunk 8]
```diff
@@ -625,5 +630,53 @@ def rms_norm_fn(
...
+        x: Tensor,
+        weight: Optional[Tensor],
+        bias: Optional[Tensor],
+        eps: float,
+        is_rms_norm: bool = False,
+        out: Optional[Tensor] = None,
+    ):
+        return norm_infer_native(x, weight, bias, eps, is_rms_norm, out)
+
+    @maybe_wrap_jit_kernel_debug(op_name="jit_kernel.diffusion.triton.rms_norm_fn")
```
Comment (merrymercy) [inline python/sglang/jit_kernel/diffusion/triton/norm.py]: why do we need to manually give it a name? It should be able to auto infer the op name

File: python/sglang/jit_kernel/diffusion/triton/scale_shift.py [hunk 9]
```diff
@@ -724,6 +734,9 @@ def fuse_layernorm_scale_shift_gate_select01_kernel(
     return output, gate_out
 
 
+@maybe_wrap_jit_kernel_debug(
+    op_name="jit_kernel.diffusion.triton.fuse_residual_layernorm_scale_shift_gate_select01_kernel"
```
Comment (merrymercy) [inline python/sglang/jit_kernel/diffusion/triton/scale_shift.py]: this is too tedious. The decorator should auto infer the name

File: python/sglang/srt/models/minimax_m2.py [hunk 10]
```diff
@@ -158,6 +159,7 @@ def rmsnorm_apply_kernel_serial(
     tl.store(out2_row + offsets2, out2, mask=mask2)
 
 
+@debug_kernel_api(op_name="MiniMaxM2.rms_sumsq_serial")
```
Comment (merrymercy) [inline python/sglang/srt/models/minimax_m2.py]: it should auto infer.
The decorator can look at the file name and function name.

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: .claude/skills/debug-cuda-crash/SKILL.md (hunk 7 above)
Comment (BBuf) [reply .claude/skills/debug-cuda-crash/SKILL.md]: Done in https://github.com/sgl-project/sglang/pull/20910/changes/42beeb24a6ed0c2bda0f370794da52a14c014608

File: python/sglang/jit_kernel/diffusion/triton/norm.py (hunk 8 above)
Comment (BBuf) [reply python/sglang/jit_kernel/diffusion/triton/norm.py]: Done

File: python/sglang/jit_kernel/diffusion/triton/scale_shift.py (hunk 9 above)
Comment (BBuf) [reply python/sglang/jit_kernel/diffusion/triton/scale_shift.py]: Done

File: python/sglang/srt/models/minimax_m2.py (hunk 10 above)
Comment (BBuf) [reply python/sglang/srt/models/minimax_m2.py]: Done

**Review by merrymercy — APPROVED** [review]
